snowflake.go

```diff
@@ -0,0 +1,341 @@
+	maxRolenameChunkLen      = 32 // arbitrarily chosen
+	defaultSnowflakeRenewSQL = `
+alter user {{name}} set DAYS_TO_EXPIRY = {{expiration}};
+`
+	defaultSnowflakeRotateCredsSQL = `
+alter user {{name}} set PASSWORD = '{{password}}';
+`
+	defaultSnowflakeDeleteSQL = `
+drop user {{name}};
+`
+)
+
+var (
+	_ dbplugin.Database = (*SnowflakeSQL)(nil)
+)
+
+func New() (interface{}, error) {
+	db := new()
+	// Wrap the plugin with middleware to sanitize errors
+	dbType := dbplugin.NewDatabaseErrorSanitizerMiddleware(db, db.secretValues)
+	return dbType, nil
+}
+
+func (s *SnowflakeSQL) secretValues() map[string]string {
+	return map[string]string{
+		s.Password: "[password]",
+	}
+}
+
+func new() *SnowflakeSQL {
+	connProducer := &connutil.SQLConnectionProducer{}
+	connProducer.Type = snowflakeSQLTypeName
+
+	db := &SnowflakeSQL{
+		SQLConnectionProducer: connProducer,
+	}
+
+	return db
+}
+
+type SnowflakeSQL struct {
+	*connutil.SQLConnectionProducer
+}
+
+func (s *SnowflakeSQL) Type() (string, error) {
+	return snowflakeSQLTypeName, nil
+}
+
+func (s *SnowflakeSQL) getConnection(ctx context.Context) (*sql.DB, error) {
+	db, err := s.Connection(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return db.(*sql.DB), nil
+}
+
+func (s *SnowflakeSQL) Initialize(ctx context.Context, req dbplugin.InitializeRequest) (dbplugin.InitializeResponse, error) {
+	err := s.SQLConnectionProducer.Initialize(ctx, req.Config, req.VerifyConnection)
+	if err != nil {
+		return dbplugin.InitializeResponse{}, err
+	}
+
+	resp := dbplugin.InitializeResponse{
+		Config: req.Config,
+	}
+
+	return resp, nil
+}
+
+func (s *SnowflakeSQL) NewUser(ctx context.Context, req dbplugin.NewUserRequest) (dbplugin.NewUserResponse, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	statements := req.Statements.Commands
+	if len(statements) == 0 {
+		return dbplugin.NewUserResponse{}, dbutil.ErrEmptyCreationStatement
+	}
+
+	username, err := s.generateUsername(req)
+	if err != nil {
+		return dbplugin.NewUserResponse{}, err
+	}
+
+	password := req.Password
+	expirationStr, err := calculateExpirationString(req.Expiration)
+	if err != nil {
+		return dbplugin.NewUserResponse{}, err
+	}
+
+	// Get the connection
+	db, err := s.getConnection(ctx)
+	if err != nil {
+		return dbplugin.NewUserResponse{}, err
+	}
+
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return dbplugin.NewUserResponse{}, err
+	}
+	defer tx.Rollback()
+
+	// Execute each query
+	for _, stmt := range statements {
+		// it's fine to split the statements on the semicolon.
+		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
+			query = strings.TrimSpace(query)
+			if len(query) == 0 {
+				continue
+			}
+
+			m := map[string]string{
+				"name":       username,
+				"username":   username,
+				"password":   password,
+				"expiration": expirationStr,
+			}
+
+			if err := dbtxn.ExecuteTxQuery(ctx, tx, m, query); err != nil {
+				return dbplugin.NewUserResponse{}, err
+			}
+		}
+	}
+
```

Comment:
I think this is still missing a `tx.Commit()`?

Reply:
:(